protocols/1624/csv_dilution.ot2.py

```diff
@@ -0,0 +1,189 @@
+}
+
+# create custom labware
+custom_tips_name = 'custom-tips-300ul'
+if custom_tips_name not in labware.list():
+    labware.create(
+        custom_tips_name,
+        grid=(12, 8),
+        spacing=(9, 9),
+        diameter=3,
+        depth=60
+        )
+
+deepwell_name = 'Greiner-Masterblock-2ml'
+if deepwell_name not in labware.list():
+    labware.create(
+        deepwell_name,
+        grid=(12, 8),
+        spacing=(9, 9),
+        diameter=8.2,
+        depth=40,
+        volume=2000
+    )
+
+dil_plate_name = 'Nunc-96-well'
+if dil_plate_name not in labware.list():
+    labware.create(
+        dil_plate_name,
+        grid=(12, 8),
+        spacing=(9, 9),
+        diameter=6.7,
+        depth=9.8,
+        volume=450
+    )
+
+res_name = 'Agilent-1-well-reservoir'
+if res_name not in labware.list():
+    labware.create(
+        res_name,
+        grid=(1, 1),
+        spacing=(0, 0),
+        diameter=72,
+        depth=39.22,
+        volume=290000
+    )
+
+# load labware
+source_plate = labware.load(deepwell_name, '1', 'source plate')
+dil_plates = [labware.load(
+                  dil_plate_name,
+                  slot,
+                  'intermediate dilution plate ' + str(i+1)
+              ) for i, slot in enumerate(['4', '5', '2'])]
+end_plate = labware.load(dil_plate_name, '3', 'end plate')
+h2o_res = labware.load(res_name, '6', 'H2O reservoir')
+hcl_res = labware.load(res_name, '7', 'HCl reservoir')
+tips = [labware.load(custom_tips_name, slot) for slot in ['8', '9']]
+
+# reagents
+h2o = h2o_res.wells('A1')
+hcl = hcl_res.wells('A1')
+
+# tip setup
+h2o_tip = tips[0].wells('H12')
+
+example_csv = """,,,,,,,,,,,,,,,,,
+,Dilution:,,,,,,,,,,,,,,,20,200-500 µL
+,,,,,,,,,,,,,,,,100,500 -1000µL
+,,,,,,,,,,,,,,,,200,"1 mL - 1,5 mL"
+,"Final Dilution is performed in 0,1 HCl",,,,,,,,,,,,,,,400,">1,5 mL"
+,,,,,,,,,,,,,,,,Empty,
+,dilution:,,,,,,,,,,,,,,,,
+,,,,,,,,,,,,,,,,,
+,,1,2,2,3,4,5,6,7,8,9,10,11,12,,,
+,A,20,400,Empty,20,20,20,20,20,20,20,20,20,20,,,
+,B,20,400,Empty,100,20,100,100,100,20,20,20,Empty,Empty,,,
+,C,20,400,Empty,100,20,100,100,100,20,20,20,Empty,Empty,,,
+,D,20,400,Empty,400,20,400,400,400,20,20,20,400,400,,,
+,E,20,200,Empty,400,20,400,400,400,20,20,20,400,400,,,
+,F,20,400,Empty,Empty,20,Empty,Empty,Empty,20,20,20,200,200,,,
+,G,20,200,Empty,Empty,20,Empty,Empty,Empty,20,20,20,200,200,,,
+,,,,,,,,,,,,,,,,,
+"""
+
+
+def run_custom_protocol(
+        csv_file: FileInput = example_csv,
+        pipette_mount: StringSelection('right', 'left') = 'right',
+        dilution_total_mixing_volume_in_ul: float = 300
+):
+
+    # volume check
+    if (
+            dilution_total_mixing_volume_in_ul > 300 or
+            dilution_total_mixing_volume_in_ul < 30
+    ):
+        raise Exception('Invalid dilution total mixing volume.')
+
+    # pipettes
+    p300 = instruments.P300_Single(mount=pipette_mount, tip_racks=tips)
+
+    # parse CSV
+    dil_lines = csv_file.splitlines()[9:17]
```

Comment:
This is a very specific CSV layout - would it be easier for them to trim off the excess info? Also, to make protocols more re-usable, I think having a screenshot of the format of the required CSV would be helpful

Reply:
I think that's a good idea. I'll go ahead and reach out to him and add the screenshot to the README.